### ISSUE TITLE
Installation issue: hpl and amdblis

### ISSUE THREAD

**Issue by yellowhat** [LLNL/spack#16418]
### Steps to reproduce the issue
```console
$ spack install hpl ^amdblis
==> Error: ProcessError: Command exited with status 1:
    '/var/tmp/pbs.42.ip-0B010004/vgervasi/spack-stage/spack-stage-hpl-2.3-ivrdrdzm7idjaivwvx7ttt7rbekcxveb/spack-src/configure' '--prefix=/shared/home/vgervasi/spack/opt/spack/linux-centos7-x86_64/gcc-9.2.0/hpl-2.3-ivrdrdzm7idjaivwvx7ttt7rbekcxveb' 'CFLAGS=-O3'

1 error found in build log:
     32    checking for dgemm_ in Apple VecLib... no
     33    checking for dgemm_ in IBM ESSL... no
     34    checking for dgemm_ in NVIDIA nvblas... no
     35    checking for dgemm_ in Generic BLAS... no
     36    checking for dgemm_ in OpenBLAS... checking for dgemm_ in -lopenblas
           ... no
     37    no
  >> 38    configure: error: BLAS not found

See build log for details:
  /var/tmp/pbs.42.ip-0B010004/vgervasi/spack-stage/spack-stage-hpl-2.3-ivrdrdzm7idjaivwvx7ttt7rbekcxveb/spack-build-out.txt

$ cat /var/tmp/pbs.42.ip-0B010004/vgervasi/spack-stage/spack-stage-hpl-2.3-ivrdrdzm7idjaivwvx7ttt7rbekcxveb/spack-build-out.txt
==> 34436: hpl: Executing phase: 'autoreconf'
==> 34436: hpl: Executing phase: 'configure'
==> [2020-05-01-15:20:59.939307, 34489] '/var/tmp/pbs.42.ip-0B010004/vgervasi/spack-stage/spack-stage-hpl-2.3-ivrdrdzm7idjaivwvx7ttt7rbekcxveb/spack-src/configure' '--prefix=/shared/home/vgervasi/spack/opt/spack/linux-centos7-x86_64/gcc-9.2.0/hpl-2.3-ivrdrdzm7idjaivwvx7ttt7rbekcxveb' 'CFLAGS=-O3'
checking for gcc... /shared/home/vgervasi/spack/lib/spack/env/gcc/gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether /shared/home/vgervasi/spack/lib/spack/env/gcc/gcc accepts -g... yes
checking for /shared/home/vgervasi/spack/lib/spack/env/gcc/gcc option to accept ISO C89... none needed
checking whether /shared/home/vgervasi/spack/lib/spack/env/gcc/gcc understands -c and -o together... yes
checking for function MPI_Init... no
checking for function MPI_Init in -lmpi... yes
checking for mpi.h... yes
checking for ranlib... ranlib
checking for a BSD-compatible install... /bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports the include directive... yes (GNU style)
checking whether make supports nested variables... yes
checking dependency style of /shared/home/vgervasi/spack/lib/spack/env/gcc/gcc... gcc3
checking for dgemm_ in OpenBLAS... no
checking for dgemm_ in Atlas Fortran BLAS... no
checking for dgemm_ in Sequential Intel MKL LP64 (group)... no
checking for dgemm_ in Sequential Intel MKL LP64... no
checking for dgemm_ in AMD's ACML... no
checking for dgemm_ in Accelerate... no
checking for dgemm_ in Apple VecLib... no
checking for dgemm_ in IBM ESSL... no
checking for dgemm_ in NVIDIA nvblas... no
checking for dgemm_ in Generic BLAS... no
checking for dgemm_ in OpenBLAS... checking for dgemm_ in -lopenblas... no
no
configure: error: BLAS not found
```

Is a HPL or amdbliss error?

Thanks

**Comment by adamjstewart** [comment]
Can you share your `Make.<arch>` file from the HPL installation? That will specify the flags used to search for BLAS. Also, does the installation work with `blis`?

**Comment by yellowhat** [comment]
In the spack-src folder there is only `Make.top`, is this the you are looking for?

**Comment by adamjstewart** [comment]
Oh, nvm, apparently that was for old versions of HPL (2.2 and below). Can you try changing:
```python
        if (self.spec.satisfies('^intel-mkl') or                                
            self.spec.satisfies('^intel-parallel-studio+mkl')):                 
            config.append('LDFLAGS={0}'.format(                                 
                self.spec['blas'].libs.ld_flags))                               
                                                                                
        return config    
```
to remove the if-statement and always add those flags?

**Comment by yellowhat** [comment]
Just tried still same error.

**Comment by jlow2016** [comment]
I can get it working by editing the configure script to check for Generic BLAS with the Spack BLAS ld_flags.
I've added the line in the hpl package.py file: 

```
    @when('@2.3:')
    def configure_args(self):
        filter_file(r'^libs10=.*', "libs10=%s" % self.spec['blas'].libs.ld_flags, 'configure')

        config = [
            'CFLAGS=-O3'
        ]
```

**Comment by yellowhat** [comment]
> I can get it working by editing the configure script to check for Generic BLAS with the Spack BLAS ld_flags.
> I've added the line in the hpl package.py file:
> 
> ```
>     @when('@2.3:')
>     def configure_args(self):
>         filter_file(r'^libs10=.*', "libs10=%s" % self.spec['blas'].libs.ld_flags, 'configure')
> 
>         config = [
>             'CFLAGS=-O3'
>         ]
> ```

I have just tried:
```

    @when('@2.3:')
    def configure_args(self):
        filter_file(r'^libs10=.*', "libs10=%s" % self.spec['blas'].libs.ld_flags, 'configure')

        config = [
            'CFLAGS=-O3'
        ]

#        if (self.spec.satisfies('^intel-mkl') or
#            self.spec.satisfies('^intel-parallel-studio+mkl')):
        config.append('LDFLAGS={0}'.format(self.spec['blas'].libs.ld_flags))

        return config
```
but I am still getting the same error.

**Comment by jlow2016** [comment]
Not sure why it's not working for you.

My config.log contains:

```
configure:4184: checking for dgemm_ in Generic BLAS
configure:4206: /home/jonathan/Library/HPC/spack/lib/spack/env/gcc/gcc -o conftest -O3   conftest.c -L/home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/amdblis-2.1-ark5gaduve5vxztxzpn2k4x4apnvblri/lib -lblis >&5
configure:4206: $? = 0
configure:4213: result: yes
configure:4282: checking for dgemm_
configure:4282: /home/jonathan/Library/HPC/spack/lib/spack/env/gcc/gcc -o conftest -O3   conftest.c -L/home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/amdblis-2.1-ark5gaduve5vxztxzpn2k4x4apnvblri/lib -lblis -lmpi  >&5
configure:4282: $? = 0
configure:4282: result: yes
```

with the spack-build-out.txt containing:

```
checking for dgemm_ in OpenBLAS... no
checking for dgemm_ in Atlas Fortran BLAS... no
checking for dgemm_ in Sequential Intel MKL LP64 (group)... no
checking for dgemm_ in Sequential Intel MKL LP64... no
checking for dgemm_ in AMD's ACML... no
checking for dgemm_ in Accelerate... no
checking for dgemm_ in Apple VecLib... no
checking for dgemm_ in IBM ESSL... no
checking for dgemm_ in NVIDIA nvblas... no
checking for dgemm_ in Generic BLAS... yes
checking for dgemm_... yes
```

By the way, I added just the additional line to the original hpl package file:

```
    @when('@2.3:')
    def configure_args(self):
        filter_file(r'^libs10=.*', "libs10=%s" % self.spec['blas'].libs.ld_flags,  'configure')

        config = [
            'CFLAGS=-O3'
        ]

        if (self.spec.satisfies('^intel-mkl') or
            self.spec.satisfies('^intel-parallel-studio+mkl')):
            config.append('LDFLAGS={0}'.format(
                self.spec['blas'].libs.ld_flags))

        return config
```

**Comment by jlow2016** [comment]
[spack-build-out.txt](https://github.com/spack/spack/files/4576199/spack-build-out.txt)
[config.log](https://github.com/spack/spack/files/4576201/config.log)

Full logfiles here.

**Comment by yellowhat** [comment]
May I ask how did you install hpl? spack install hpl ^amdblis? Where did you get the `config.log` file?

Thanks

**Comment by jlow2016** [comment]
```
jonathan@N131WU:~/Library/HPC$ spack spec -Il hpl ^amdblis
Input spec
--------------------------------
 -   hpl
 -       ^amdblis

Concretized
--------------------------------
 -   dte6kkb  hpl@2.3%gcc@7.5.0~openmp arch=linux-ubuntu18.04-skylake
[+]  ark5gad      ^amdblis@2.1%gcc@7.5.0+blas+cblas+shared+static threads=none arch=linux-ubuntu18.04-skylake
[+]  aqe4ejm          ^python@3.7.7%gcc@7.5.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=linux-ubuntu18.04-skylake
[+]  yjwcvti              ^bzip2@1.0.8%gcc@7.5.0+shared arch=linux-ubuntu18.04-skylake
[+]  624pjac                  ^diffutils@3.7%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  7ddqjmf                      ^libiconv@1.16%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  luqxdp4              ^expat@2.2.9%gcc@7.5.0+libbsd arch=linux-ubuntu18.04-skylake
[+]  gvadyz7                  ^libbsd@0.10.0%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  a3onxbk              ^gdbm@1.18.1%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  l3bzhbk                  ^readline@8.0%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  nlvmejf                      ^ncurses@6.2%gcc@7.5.0~symlinks+termlib arch=linux-ubuntu18.04-skylake
[+]  uhvjvnw                          ^pkgconf@1.6.3%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  o7lwh5y              ^gettext@0.20.2%gcc@7.5.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-skylake
[+]  dye3m4s                  ^libxml2@2.9.10%gcc@7.5.0~python arch=linux-ubuntu18.04-skylake
[+]  jkbd22l                      ^xz@5.2.5%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  fz2bs56                      ^zlib@1.2.11%gcc@7.5.0+optimize+pic+shared arch=linux-ubuntu18.04-skylake
[+]  2gudv7n                  ^tar@1.32%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  b72i3sc              ^libffi@3.3%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  y5ou4xi              ^openssl@1.1.1g%gcc@7.5.0+systemcerts arch=linux-ubuntu18.04-skylake
[+]  32yrykj                  ^perl@5.30.2%gcc@7.5.0+cpanm+shared+threads arch=linux-ubuntu18.04-skylake
[+]  zj656na              ^sqlite@3.31.1%gcc@7.5.0+column_metadata+fts~functions~rtree arch=linux-ubuntu18.04-skylake
[+]  zqtqypv      ^openmpi@3.1.5%gcc@7.5.0~atomics~cuda~cxx~cxx_exceptions fabrics=none +gpfs~java~legacylaunchers~memchecker~pmi+runpath schedulers=none ~sqlite3+static~thread_multiple+vt arch=linux-ubuntu18.04-skylake
[+]  2as2by2          ^hwloc@1.11.11%gcc@7.5.0~cairo~cuda~gl+libxml2~netloc~nvml+pci+shared arch=linux-ubuntu18.04-skylake
[+]  xehyegf              ^libpciaccess@0.13.5%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  oljw33w                  ^libtool@2.4.6%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  erqdtam                      ^m4@1.4.18%gcc@7.5.0 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-ubuntu18.04-skylake
[+]  aymspcl                          ^libsigsegv@2.12%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  se2a2e7                  ^util-macros@1.19.1%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  zn7xa6m              ^numactl@2.0.12%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  fezkgv3                  ^autoconf@2.69%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
[+]  xbmzejy                  ^automake@1.16.2%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
```
```
jonathan@N131WU:~/Library/HPC$ spack install hpl ^amdblis
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libiconv-1.16-7ddqjmff6zg3zer572yuqnvzk2ebrkfc
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libbsd-0.10.0-gvadyz7pk4n6zmbn2kepfwjgdfc6yr2x
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/pkgconf-1.6.3-uhvjvnwuzaug3giyrmwky6ylaqkvmrud
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/xz-5.2.5-jkbd22lgypfj3n4amfpfpgveyjm53vwi
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/zlib-1.2.11-fz2bs562jhc2spgubs3fvq25g3qymz6x
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libffi-3.3-b72i3sc7ldo3izhuta3voe4akyvc52cn
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libsigsegv-2.12-aymspclv5aqdkwqkxvlxxerrimgsjzd2
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/util-macros-1.19.1-se2a2e74oyusj2r4esgcb7pr3qhh45ef
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/tar-1.32-2gudv7nnasphqr6zwrvoulakgqkfnbha
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/diffutils-3.7-624pjac4maorpjwa4kl7ydasl5jy7wgy
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/expat-2.2.9-luqxdp4jubd4yp3xoypxgzv6ai3ev23h
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/ncurses-6.2-nlvmejfn4n7jkkot2nglflrtwpu56o27
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libxml2-2.9.10-dye3m4ss7rgzc7ikjfs2jegond77vu25
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/m4-1.4.18-erqdtam4yfnuwhwzmk2n5u4wbwcb5wkv
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/bzip2-1.0.8-yjwcvti7ehjkvqcu67rdzymdawi6nruk
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/readline-8.0-l3bzhbkekxg3ggeov4cjh3gaj5jdplet
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libtool-2.4.6-oljw33wi4jixtfxnwe56rijcmgw3t4o4
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/gettext-0.20.2-o7lwh5yphan7a7w5ooiwmy7o4olcytpy
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/gdbm-1.18.1-a3onxbk63mvypoeocklamcj3q7jcyfyj
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/sqlite-3.31.1-zj656naoe5ivo46bafjxex6ipi6fg736
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/libpciaccess-0.13.5-xehyegf2mqjdm2spqcwmp7ybax2sspwu
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/perl-5.30.2-32yrykjqfsxadecvxz7jr2jews5larqi
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/autoconf-2.69-fezkgv3ggnzx752r3zrufvmxr2rmc3vb
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/openssl-1.1.1g-y5ou4xinlxtxiofan6xxmzfgt36rf6jb
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/automake-1.16.2-xbmzejywpd4yug2pwwv4jrcsva4vcs5f
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/python-3.7.7-aqe4ejm7wwvogbkwyfqdpmudbpijaus6
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/numactl-2.0.12-zn7xa6mvdspq7ypjhetd66yz5dfdrtsk
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/amdblis-2.1-ark5gaduve5vxztxzpn2k4x4apnvblri
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/hwloc-1.11.11-2as2by2fq7w2lvckvhqz5bzgxzrygjen
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/openmpi-3.1.5-zqtqypvs3vpmbs4wvwod2tmmz2eknvde
==> 27633: Installing hpl
==> Using cached archive: /home/jonathan/Library/HPC/spack/var/spack/cache/_source-cache/archive/32/32c5c17d22330e6f2337b681aded51637fb6008d3f0eb7c277b163fadd612830.tar.gz
==> Staging archive: /tmp/jonathan/spack-stage/spack-stage-hpl-2.3-dte6kkb7xqbsv3m5jfdvnn6kk36r3z5j/hpl-2.3.tar.gz
==> Created stage in /tmp/jonathan/spack-stage/spack-stage-hpl-2.3-dte6kkb7xqbsv3m5jfdvnn6kk36r3z5j
==> No patches needed for hpl
==> 27633: hpl: Building hpl [AutotoolsPackage]
==> 27633: hpl: Executing phase: 'autoreconf'
==> 27633: hpl: Executing phase: 'configure'
==> 27633: hpl: Executing phase: 'build'
==> 27633: hpl: Executing phase: 'install'
==> 27633: hpl: Successfully installed hpl
  Fetch: 0.00s.  Build: 12.35s.  Total: 12.36s.
[+] /home/jonathan/Library/HPC/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-7.5.0/hpl-2.3-dte6kkb7xqbsv3m5jfdvnn6kk36r3z5j
```
The config.log file is \<stage-directory\>/spack-src/config.log where the stage directory is the line with "Created stage in" (assuming the install fails, otherwise it is cleared by default after successful install).

I'm using the git develop branch:
```
jonathan@N131WU:~/Library/HPC/spack$ git show
commit 87f78137a54b2d317172e18c1d46a46c31362c6e (HEAD -> develop, origin/develop, origin/HEAD)
Author: Axel Huebl <axel.huebl@plasma.ninja>
Date:   Mon May 4 00:39:20 2020 -0700
```

**Comment by adamjstewart** [comment]
> (assuming the install fails, otherwise it is cleared by default after successful install).

You can add `--keep-stage` to prevent the stage directory from being removed. Also, the `config.log` should be copied to `<prefix>/.spack/config.log` upon successful installation.

**Comment by yellowhat** [comment]
If I patch and run:
```
spack install hpl +openmp ^amdblis
```
it compiles, instead:
```
spack install hpl +openmp ^amdblis threads=openmp
```
it stil returns the same error.

Any ideas?

**Comment by jlow2016** [comment]
I think this is related to pull request https://github.com/spack/spack/pull/15566

You'll need to patch the blis package to look for the libblis-mt.so library

```
    @property
    def libs(self):
        return find_libraries(['libblis', 'libblis-mt'], root=self.prefix, recursive=True)
```

**Comment by yellowhat** [comment]
Thank you very much, that fixed the issues.

**Comment by adamjstewart** [comment]
Want to submit a PR?

**Comment by yellowhat** [comment]
I am trying to compile hpl with "+openmp":
```
spack install hpl +openmp ^openmpi ^amdblis threads=openmp arch=linux-centos7-zen2
```
But if I try to run it no multiple threads are spawn for each MPI rank.

I can see that "OMP_DEFS" is present in setup/Make.Linux_Intel64, but I do not see it in `config.log`.
How can I check which `Make.*` is used during the build?

Thanks

**Comment by yellowhat** [comment]
Yes how do I do it?

**Comment by adamjstewart** [comment]
Our [Contribution Guide](https://spack.readthedocs.io/en/latest/contribution_guide.html) has some tips. Basically, create a new git branch, add and commit the change, and then push.

**Comment by adamjstewart** [comment]
Closed by #16476
